### PR TITLE
Wrap histogram

### LINKED ISSUE #563: Wrap histogram
**Description of the desired feature**

The **histogram** module is useful and should be wrapped.

Some useful resources:

- [GMT's histogram manpage](https://docs.generic-mapping-tools.org/dev/histogram.html)
- [GMT.jl's histogram wrapper](https://www.generic-mapping-tools.org/GMT.jl/latest/#GMT.histogram)
- [numpy.histogram](https://numpy.org/doc/stable/reference/generated/numpy.histogram.html)
- [matplotlib.pyplot.hist](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.hist.html)

**Are you willing to help implement and maintain this feature?** 
Yes, but may have no time recently, so contributions are welcomed.

### PR DESCRIPTION
**Pull request by willschlitzer**
Wrapping the [histogram](https://docs.generic-mapping-tools.org/6.1/histogram.html) module

**Preview** at https://pygmt-git-wrap-histogram-gmt.vercel.app/api/generated/pygmt.Figure.histogram.html

Fixes #563

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [x] Add new public functions/methods/classes to `doc/api/index.rst`.
- [x] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by willschlitzer** [comment]
I'm not sure how to use the virtualfile to import a table. I'm trying:
`table = [1,1,1,1,1,1,2,2,2,3,4,5,6,7,8,8,8,8,8,8]`
`np_table = np.array(table)`
`fig.histogram(table=np_table, projection="X10c/25c", T=1, G="green")`

But I get `TypeError: histogram() got multiple values for argument 'table'`

I'm not sure how to fix this, as it seems like I'm passing the correct input in (but clearly not).Any ideas @GenericMappingTools/python ?

File: pygmt/src/histogram.py
Comment (seisman) [inline pygmt/src/histogram.py:21]: I believe you forgot `self`:
```suggestion
def histogram(self, table, **kwargs):
```

**Comment by seisman** [comment]
I would suggest you only focus on the plotting part in this PR, and implement the inquire (**-I**) feature in a separate PR (after some discussions as in #896).

File: pygmt/src/histogram.py
Comment (willschlitzer) [reply pygmt/src/histogram.py]: Yep, I copied directly from grdinfo to get the virtual file example and forgot to include self from the other plotting functions. 🤦🏻‍♂️

**Comment by weiji14** [comment]
> I would suggest you only focus on the plotting part in this PR, and implement the inquire (**-I**) feature in a separate PR (after some discussions as in #896).

Just on this point (and as I've mentioned to Jiayuan at https://github.com/GenericMappingTools/pygmt/pull/1145#pullrequestreview-628868487), I'd recommend wrapping as few aliases as you can get away with in this PR. We can open up separate PRs to wrap any complicated parameter like `inquire=I`, or `transparency=t`, and the gallery example/tutorial can be left for another day too.

To be clear, that means you just have to alias the required arguments (https://docs.generic-mapping-tools.org/dev/histogram.html#required-arguments), and the common aliases (-R, -J, -c, -p, -t, etc) to pass. Maybe add one minimal unit test to hit the code coverage target. I see you've done a few aliases already (and you can leave them in if you want), but don't feel like you have to complete all of them.

Example workflow would be to do: Minimal PR (this one) -> Gallery Example PR -> Complete documentation/aliases PR -> Full Tutorial PR

P.S. Yes, we should document this 'policy' about wrapping modules chunk by chunk somewhere, but thought I'd mention it to you first.

**Comment by willschlitzer** [comment]
> > I would suggest you only focus on the plotting part in this PR, and implement the inquire (**-I**) feature in a separate PR (after some discussions as in #896).
> 
> Just on this point (and as I've mentioned to Jiayuan at [#1145 (review)](https://github.com/GenericMappingTools/pygmt/pull/1145#pullrequestreview-628868487)), I'd recommend wrapping as few aliases as you can get away with in this PR. We can open up separate PRs to wrap any complicated parameter like `inquire=I`, or `transparency=t`, and the gallery example/tutorial can be left for another day too.
> 
> To be clear, that means you just have to alias the required arguments (https://docs.generic-mapping-tools.org/dev/histogram.html#required-arguments), and the common aliases (-R, -J, -c, -p, -t, etc) to pass. Maybe add one minimal unit test to hit the code coverage target. I see you've done a few aliases already (and you can leave them in if you want), but don't feel like you have to complete all of them.
> 
> Example workflow would be to do: Minimal PR (this one) -> Gallery Example PR -> Complete documentation/aliases PR -> Full Tutorial PR
> 
> P.S. Yes, we should document this 'policy' about wrapping modules chunk by chunk somewhere, but thought I'd mention it to you first.

I'm on board with with suggested workflow. I think it keeps the pull request more manageable by not including everything under one review,

File: pygmt/src/histogram.py
Comment (willschlitzer) [reply pygmt/src/histogram.py]: @seisman The use of `self` is causing a style check failure because it isn't used in the function; should I disable pylint for this error (follow-up question: how would I do that?)? The function itself appears to be working correctly, but needs `self` for it to work.

**Comment by willschlitzer** [comment]
Screenshot from 2021-04-09 07-12-21

The parameters list is coming across like this, and I'm not sure how to fix it. As far as I can tell, the docstring format for `histogram` is the same as other functions, but there is some difference that is causing the parameters to list as indented bullet points.

**Comment by meghanrjones** [comment]
> Screenshot from 2021-04-09 07-12-21
> 
> The parameters list is coming across like this, and I'm not sure how to fix it. As far as I can tell, the docstring format for `histogram` is the same as other functions, but there is some difference that is causing the parameters to list as indented bullet points.

This is because of the formatting updates with the latest version of the sphinx_rtd_theme, added in https://github.com/GenericMappingTools/gmt/pull/5072. I will try to fix this as part of #1151.

File: pygmt/src/histogram.py
Comment (weiji14) [inline pygmt/src/histogram.py:55]: Need to add `self.preprocess` to ensure histogram is plotted to the right figure always.

```suggestion
    """
    kwargs = self._preprocess(**kwargs)  # pylint: disable=protected-access
```

File: pygmt/src/histogram.py
Comment (weiji14) [inline pygmt/src/histogram.py:63]: Err, I don't think there's any data output from `histogram` right now since the inquire (-I) mode doesn't work yet (still need to discuss at #896). So maybe take this out (also need to remove the `GMTTempFile` import and `with` statement above).

```suggestion
```

**Review by weiji14 — COMMENTED** [review]
Indentation issue seems to be fixed in the live API docs at https://pygmt-git-wrap-histogram-gmt.vercel.app/api/generated/pygmt.Figure.histogram (thanks Meghan). So here comes the comments!

File: pygmt/src/histogram.py
Comment (willschlitzer) [reply pygmt/src/histogram.py:22]: Works for me!

File: pygmt/src/histogram.py
Comment (willschlitzer) [reply pygmt/src/histogram.py:58]: @weiji14 Can you explain what this does? I'm committing it, but not sure what this accomplishes.

File: pygmt/src/histogram.py
Comment (willschlitzer) [reply pygmt/src/histogram.py:69]: Sounds good, I'll make this change next time I'm on my home computer.

Comment (willschlitzer) [reply pygmt/src/histogram.py]: @weiji14 fixed in d9b8e39

File: pygmt/src/histogram.py
Comment (seisman) [inline pygmt/src/histogram.py:23]: Rename the parameter to `data`, similar to other functions?
```suggestion
def histogram(self, data, **kwargs):
```

File: pygmt/src/histogram.py
Comment (seisman) [inline pygmt/src/histogram.py:16]: Is `series` a good name? Other Python histogram functions (e.g., [`numpy.histogram`](https://numpy.org/doc/stable/reference/generated/numpy.histogram.html), [`plt.histogram`](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.hist.html) use `bins` instead.

File: pygmt/src/histogram.py
Comment (weiji14) [reply pygmt/src/histogram.py:23]: Some functions use data, some use table, I'd actually prefer table since histogram isn't a geographical plot.

File: pygmt/src/histogram.py
Comment (weiji14) [reply pygmt/src/histogram.py:16]: See my comment on https://github.com/GenericMappingTools/pygmt/pull/1072#discussion_r611029660. Probably not a good idea to use `bins` as the input for `-T` is structured quite differently to the `bins` argument in `np.histogram` and `plt.histogram`.

Comment (seisman) [reply pygmt/src/histogram.py:16]: Actually, it depends on how we want to implement the **-T** option in PyGMT.

histogram **-T** have following usages:

1. -T6
2. -T0/200/5
3. -Tfile
4. -T0,3,4,6,10

The current implementation doesn't work well for usage 4, since users have to pass a string "0,3,4,6,10". 

Instead, if we adopt the syntax of `np.histogram`, we can use more Pythonic syntax
```
bins=6
bins=np.arange(0, 200, 5)
bins="file"
bins=[0, 3, 4, 6, 10)
```

Comment (weiji14) [reply pygmt/src/histogram.py:16]: Currently we're using `@kwargs_to_strings(T="sequence")` which implies Option 2 (-T0/200/5) and I think this works well for `makecpt`/`grd2cpt` where regular sized bins are the norm, and also `histogram` to some extent. Doing Option 4 would mean using `@kwargs_to_strings(T="sequence_comma")` and I guess this would be fine for people wanting irregular sized bins for `histogram`.

I don't often plot histograms, so I'll leave it up to you all to decide which one is better. Ideally we would standardize the `-T` option across all modules, but we could also use `series` (with T="sequence") ` for `makecpt`/`grd2cpt`, and `bins` (with T="sequence_comma")  for `histogram` if that makes more sense for new users.

Comment (seisman) [reply pygmt/src/histogram.py:16]: Just a note here: we don't have to use `@use_alias` and `@kwargs_to_strings` for **-T**.

Instead, we can check if the `bins` (or any other parameter name) argument is a string, one value, three values, or a list of values. I think it provides the most Pythonic syntax and also the greatest flexibility.

File: pygmt/src/histogram.py
Comment (willschlitzer) [reply pygmt/src/histogram.py:23]: For the sake of consistency with the GMT terminology, my vote is it for to remain `table`

File: pygmt/src/histogram.py
Comment (willschlitzer) [reply pygmt/src/histogram.py:16]: I'm not looking to shirk responsibility for wrapping `histogram`, but are many geoscientists creating histograms with GMT? I think it's good to provide the capability in PyGMT in case anyone is looking to incorporate it into their figures, but I imagine almost all PyGMT users will be familiar with numpy and matplotlib and using that instead for plots like this, so I don't wish to make the PyGMT syntax overly different than GMT syntax.

Comment (weiji14) [reply pygmt/src/histogram.py:16]: Good point. For a single histogram plot, it would be be much easier to use something like the built-in [pandas hist function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.hist.html) (e.g. just do `df.hist()`). But for say, mixing a `grdview` plot with a `histogram` as an 'inset', then using PyGMT would be better :wink:

Let's get another opinion. Ping @core-man since he seemed interested in `histogram` at https://github.com/GenericMappingTools/pygmt/issues/563#issuecomment-801246659. What syntax do you think PyGMT's `histogram` should follow, GMT's or Matplotlib's?

Comment (core-man) [reply pygmt/src/histogram.py:16]: During my 10-year seismological studies, I only use GMT to plot figures and sometimes use AI/3D max to plot cartoons. I began to use Python and PyGMT at the same time last year, so I have never learned matplotlib. So for my study, GMT/PyGMT is enough right now. For the PyGMT scope I understand now, we'd like to provide a good Python interface for GMT. So as we go to v1.0.0, I prefer more Pythonic syntax instead of strict GMT's. At that time, I think (also hope) most GMT users will learn GMT by using PyGMT.

For this issue, I think the current implementation is okay and most users can use it well (I never use option 4). However, it's better we could refractor all the similar parameters to be more Pythonic syntax as we go to v1.0.0, e.g.,
```
bins=6
bins=np.arange(0, 200, 5)
bins="file"
bins=[0, 3, 4, 6, 10)
```

Comment (willschlitzer) [reply pygmt/src/histogram.py:16]: I agree that we should move to a more Pythonic syntax overall. Assuming that is the eventual plan, it's creating more work for our future selves by wrapping new features with a GMT syntax. That being said, I think there are more commonly-used features (setting map projections, updating pen/fill settings, adding a frame) that should be prioritized if the plan is to begin incorporating Pythonic syntax. This is certainly an issue outside of the scope of wrapping `histogram`, but I think if we are changing new features to be in-line with a more Pythonic syntax we should be also updating our old features ASAP.

Comment (willschlitzer) [reply pygmt/src/histogram.py:16]: Maybe I am missing it, but is there a consensus on how this function should be modified?

File: pygmt/src/histogram.py
Comment (willschlitzer) [reply pygmt/src/histogram.py:23]: @seisman Do you have an opinion on this? Is it fine if I leave it as `table`?

File: pygmt/src/histogram.py
Comment (weiji14) [reply pygmt/src/histogram.py:16]: For the sake of time and consistency with other PyGMT modules like makecpt/grd2cpt, I say we stick with using `series` (-T) instead of `bins`. For better or worse, we are still heavily skewed towards the style of GMT core rather than NumPy, and decisions of what is 'Pythonic' should go in a separate issue. It is better to be consistently like GMT, rather than confuse users with a mix of NumPy-like and GMT command-line-like syntax. So @willschlitzer, let's keep things as it is for now :)

File: pygmt/src/histogram.py
Comment (weiji14) [inline pygmt/src/histogram.py:34]: Can remove one blank line here.
```suggestion
```

**Review by weiji14 — COMMENTED** [review]
Sorry for the wait @willschlitzer. I'm just about ready to approve this and put the final-review-call label on. If you could please update the branch, and re-create the `histogram` baseline image using GMT 6.2.0rc1 (or we can do it for you if you don't want to upgrade your conda environment yet) and we'll try and get this merged in the next few days, barring any major issues :crossed_fingers:

**Comment by willschlitzer** [comment]
> Sorry for the wait @willschlitzer. I'm just about ready to approve this and put the final-review-call label on. If you could please update the branch, and re-create the `histogram` baseline image using GMT 6.2.0rc1 (or we can do it for you if you don't want to upgrade your conda environment yet) and we'll try and get this merged in the next few days, barring any major issues crossed_fingers

@weiji14 Updated the baseline image with my new GMT 6.2.0rc1 environment (took much longer than I'm proud to admit to find out to use the current working branch of PyGMT in your conda environment; PR coming soon to add that to contributing/install guidelines).

File: pygmt/src/histogram.py
Comment (michaelgrund) [reply pygmt/src/histogram.py:23]: Would also prefer table instead of data here.

**Review by weiji14 — APPROVED** [review]
Alright, glad you managed to update to GMT 6.2.0rc1, I agree that the upgrade process could be a lot less painful (but that's for a separate issue). Anyways, time for the final review call!

File: pygmt/src/histogram.py
Comment (weiji14) [inline pygmt/src/histogram.py:49]:
```suggestion
    series : int or str or list
```

File: pygmt/tests/test_histogram.py
Comment (seisman) [inline pygmt/tests/test_histogram.py:14]: image

Is the baseline image correct? I expect to see y=6 for x=8.

BTW, please change `projection` to a smaller dimension like `X10c/5c`.

Comment (willschlitzer) [reply pygmt/tests/test_histogram.py:14]: @seisman I just tried it on both GMT 6.1.1 and 6.2.0rc1, with the figure coming out the same way. Any idea what might be causing the miscount for 8?

Also, changing the projection to `X3c/10c` to give it a larger vertical scale; does that work?

Comment (seisman) [reply pygmt/tests/test_histogram.py:14]: It looks like an upstream bug.

Comment (willschlitzer) [reply pygmt/tests/test_histogram.py:14]: I'll try and recreate it in GMT and open an upstream issue

Comment (seisman) [reply pygmt/tests/test_histogram.py:14]: GMT automatically determines the x range from input data. In this example, xmin=1, xmax=8, but the vertical bar for x=8 is outside of the [1, 8] range, so it's not plotted.

You should be able to see the x=8 bar, if you manually set `region=[0, 9, 0, 6]`. I think we can do it temporarily so that this PR can be merged, and report it to upstream and wait for feedback and potential fix.

Comment (seisman) [reply pygmt/tests/test_histogram.py:14]: > Also, changing the projection to `X3c/10c` to give it a larger vertical scale; does that work?

BTW, the large vertical scale makes the image difficult to read, why not just use "X10c/10c" for a square basemap?

Comment (willschlitzer) [reply pygmt/tests/test_histogram.py:14]: @seisman I made your suggested changes

**Review by seisman — APPROVED** [review]

**Comment by seisman** [comment]
The "Style Checks" fail but is unrelated to changes in this PR:
```
************* Module pygmt.figure
pygmt/figure.py:82:8: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
```

**Pull request merged**
